### PR TITLE
refactor: Convert assets to TypeScript

### PR DESCRIPTION
**Pull request by bennyn**
# INFO

This PR converts all files (except `z.assets.AssetTransferState`) from the "assets" directory to TypeScript. I will convert `z.assets.AssetTransferState` in a separate PR as it is used in many different places and I want to have a solid some run first before I make that change.

### CODE REVIEW

File: src/script/assets/AssetCrypto.ts [hunk 1]
```diff
@@ -17,13 +17,11 @@
...
  */
 
-/**
- * @param {ArrayBuffer} cipherText - Encrypted plaintext
- * @param {ArrayBuffer} keyBytes - AES key used for encryption
- * @param {ArrayBuffer} referenceSha256 - SHA-256 checksum of the cipherText
- * @returns {Promise} Resolves with the decrypted asset
- */
-export const decryptAesAsset = (cipherText, keyBytes, referenceSha256) => {
+export const decryptAesAsset = async (
```
Comment (ffflorian) [inline src/script/assets/AssetCrypto.ts:26]: Please do not remove the explanations.

Comment (bennycode) [reply src/script/assets/AssetCrypto.ts:26]: You mean the comment? Which value does it provide?

Comment (ffflorian) [reply src/script/assets/AssetCrypto.ts:26]: The descriptions for the `keyBytes` and `referenceSha256` are helpful IMO. The other ones are not needed.

File: src/script/assets/AssetMetaDataBuilder.ts [hunk 2]
```diff
@@ -18,34 +18,34 @@
...
-import {chunk} from 'utils/ArrayUtil';
-import {loadFileBuffer} from 'utils/util';
-
-/**
- * Constructs corresponding asset metadata depending on the given file type
- * @param {File|Blob} file - the file to generate metadata for
- * @returns {Promise} Resolves with ImageMetaData, VideoMetaData or AudioMetaData
- */
-const buildMetadata = file => {
+import {chunk} from '../util/ArrayUtil';
```
Comment (ffflorian) [inline src/script/assets/AssetMetaDataBuilder.ts:21]:
```suggestion
import {chunk} from 'utils/ArrayUtil';
```

File: src/script/assets/AssetMetaDataBuilder.ts [hunk 3]
```diff
@@ -18,34 +18,34 @@
...
+import {capToByte, rootMeanSquare} from '../util/NumberUtil';
+import {TimeUtil} from '../util/TimeUtil';
+import {loadFileBuffer} from '../util/util';
```
Comment (ffflorian) [inline src/script/assets/AssetMetaDataBuilder.ts:24]:
```suggestion
import {loadFileBuffer} from 'utils/util';
```

File: src/script/assets/AssetMetaDataBuilder.ts [hunk 4]
```diff
@@ -18,34 +18,34 @@
...
-
-/**
- * Constructs corresponding asset metadata depending on the given file type
- * @param {File|Blob} file - the file to generate metadata for
- * @returns {Promise} Resolves with ImageMetaData, VideoMetaData or AudioMetaData
- */
-const buildMetadata = file => {
+import {chunk} from '../util/ArrayUtil';
+import {capToByte, rootMeanSquare} from '../util/NumberUtil';
+import {TimeUtil} from '../util/TimeUtil';
```
Comment (ffflorian) [inline src/script/assets/AssetMetaDataBuilder.ts:23]:
```suggestion
import {TimeUtil} from 'utils/TimeUtil';
```

File: src/script/assets/AssetMetaDataBuilder.ts [hunk 5]
```diff
@@ -18,34 +18,34 @@
...
-import {loadFileBuffer} from 'utils/util';
-
-/**
- * Constructs corresponding asset metadata depending on the given file type
- * @param {File|Blob} file - the file to generate metadata for
- * @returns {Promise} Resolves with ImageMetaData, VideoMetaData or AudioMetaData
- */
-const buildMetadata = file => {
+import {chunk} from '../util/ArrayUtil';
+import {capToByte, rootMeanSquare} from '../util/NumberUtil';
```
Comment (ffflorian) [inline src/script/assets/AssetMetaDataBuilder.ts:22]:
```suggestion
import {capToByte, rootMeanSquare} from 'utils/NumberUtil';
```

File: src/script/assets/AssetMetaDataBuilder.ts [hunk 6]
```diff
@@ -18,34 +18,34 @@
...
+import {TimeUtil} from '../util/TimeUtil';
+import {loadFileBuffer} from '../util/util';
+
+export type MetaData = Asset.AudioMetaData | Asset.VideoMetaData | Asset.ImageMetaData;
+
+const buildMetadata = (file: File | Blob): Promise<MetaData | void> => {
```
Comment (ffflorian) [inline src/script/assets/AssetMetaDataBuilder.ts:33]: The JSDoc description was pretty helpful here, too IMO.

File: src/script/assets/AssetMetaDataBuilder.ts [hunk 7]
```diff
@@ -101,15 +101,7 @@ const buildMetadataVideo = videoFile => {
...
-/**
- * Convert an error event into a plain error object.
- * This needs to be done because error events are not standardized between browser implementations.
- * @private
- * @param {Event} event - Error event
- * @returns {MediaError} Error object
- * @see https://developer.mozilla.org/en-US/docs/Web/API/Event/originalTarget
- */
-const convertEventToError = event => {
+const convertEventToError = (event: any): MediaError => {
```
Comment (ffflorian) [inline src/script/assets/AssetMetaDataBuilder.ts:117]: Same here with the description.

File: src/script/assets/AssetCrypto.ts (hunk 1 above)
Comment (bennycode) [reply src/script/assets/AssetCrypto.ts:26]: Ok! I will bring them back. 👍

File: src/script/assets/AssetRemoteData.ts [hunk 8]
```diff
@@ -17,55 +17,89 @@
  *
  */
 
-import {getLogger} from 'utils/Logger';
-import {loadUrlBuffer} from 'utils/util';
-import {getAssetUrl, setAssetUrl} from './AssetURLCache';
+import {Logger} from 'logdown';
```
Comment (ffflorian) [inline src/script/assets/AssetRemoteData.ts]: Please use `utils/Logger`.

File: src/script/util/util.js [hunk 9]
```diff
@@ -93,6 +93,12 @@ export const loadDataUrl = file => {
   });
 };
 
+/**
+ * Loads a HTTP response into an array buffer.
```
Comment (ffflorian) [inline src/script/util/util.js]:
```suggestion
 * Loads a HTTP response into an ArrayBuffer.
```

File: src/script/util/util.js [hunk 10]
```diff
@@ -93,6 +93,12 @@ export const loadDataUrl = file => {
...
+ * @param {string} url - Request URL
```
Comment (ffflorian) [inline src/script/util/util.js:98]:
```suggestion
 * @param {string} url Request URL
```

File: src/script/util/util.js [hunk 11]
```diff
@@ -93,6 +93,12 @@ export const loadDataUrl = file => {
...
+ * @param {Function} xhrAccessorFunction - Request wrapper
```
Comment (ffflorian) [inline src/script/util/util.js:99]:
```suggestion
 * @param {Function} xhrAccessorFunction Request wrapper
```

File: src/script/assets/AssetRemoteData.ts (hunk 8 above)
Comment (bennycode) [reply src/script/assets/AssetRemoteData.ts]: It won't compile. Error: TS2307.

File: src/script/assets/AssetMetaDataBuilder.ts (hunk 5 above)
Comment (bennycode) [reply src/script/assets/AssetMetaDataBuilder.ts:22]: It won't compile. Error: TS2307.

File: src/script/assets/AssetMetaDataBuilder.ts (hunk 4 above)
Comment (bennycode) [reply src/script/assets/AssetMetaDataBuilder.ts:23]: It won't compile. Error: TS2307.

File: src/script/assets/AssetMetaDataBuilder.ts (hunk 3 above)
Comment (bennycode) [reply src/script/assets/AssetMetaDataBuilder.ts:24]: It won't compile. Error: TS2307.

File: src/script/assets/AssetMetaDataBuilder.ts (hunk 2 above)
Comment (bennycode) [reply src/script/assets/AssetMetaDataBuilder.ts:21]: It won't compile. Error: TS2307.

File: src/script/assets/AssetRemoteData.ts (hunk 8 above)
Comment (ffflorian) [reply src/script/assets/AssetRemoteData.ts]: I meant: please use the `Logger` interface from '../util/Logger' (add it in line 21).

**Pull request merged**